File: backend/src/main/java/com/bsharat/triosuite/catalog/Item.java
package com.bsharat.triosuite.catalog;

import com.bsharat.triosuite.common.jpa.AuditedEntity;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import java.math.BigDecimal;
import lombok.Getter;
import lombok.Setter;

/**
 * A catalogue entry that can be added to an invoice, by search or by scanning its barcode.
 *
 * <p>{@code unitPrice} and {@code taxRate} are <em>defaults</em>. When a line is written they are
 * copied onto {@link com.bsharat.triosuite.invoice.InvoiceLine}, so re-pricing an item never
 * changes an invoice that has already been issued (ADR-0004).
 */
@Entity
@Getter
@Setter
@Table(name = "items")
public class Item extends AuditedEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 50, unique = true)
    private String sku;

    /** EAN-13 for the seeded catalogue. Nullable, because not every item is physically labelled. */
    @Column(length = 64, unique = true)
    private String barcode;

    @Column(nullable = false, length = 200)
    private String name;

    @Column(name = "unit_price", nullable = false, precision = 19, scale = 4)
    private BigDecimal unitPrice;

    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "currency_code", nullable = false)
    private Currency currency;

    @Column(name = "tax_rate", nullable = false, precision = 5, scale = 4)
    private BigDecimal taxRate;

    @Column(name = "is_active", nullable = false)
    private boolean active = true;
}

File: mobile/lib/features/invoices/domain/item.dart
import 'package:decimal/decimal.dart';

import '../../../core/money/decimal_json.dart';

/// A catalogue entry, found by search or by scanning its barcode.
///
/// [unitPrice] and [taxRate] are catalogue defaults. When the item is put on an invoice the server
/// copies them onto the line, so re-pricing an item later never changes an invoice already issued.
class Item {
  const Item({
    required this.id,
    required this.sku,
    this.barcode,
    required this.name,
    required this.unitPrice,
    required this.currencyCode,
    required this.taxRate,
    this.active = true,
  });

  final int id;
  final String sku;
  final String? barcode;
  final String name;
  final Decimal unitPrice;
  final String currencyCode;
  final Decimal taxRate;
  final bool active;

  factory Item.fromJson(Map<String, dynamic> json) {
    return Item(
      id: json['id'] as int,
      sku: json['sku'] as String,
      barcode: json['barcode'] as String?,
      name: json['name'] as String,
      unitPrice: decimalFromJson(json['unitPrice']),
      currencyCode: json['currencyCode'] as String,
      taxRate: decimalFromJson(json['taxRate']),
      active: json['active'] as bool? ?? true,
    );
  }
}
